ParsingInputFile/parser.cpp

```diff
@@ -0,0 +1,144 @@
+#include <iostream>
+#include <string>
+#include <vector>
+#include <fstream>
+#include <sstream>
+#include <unordered_map>
+#include <set>
+
+using namespace std;
+// This function reads the first character from a file and returns it.
+/**
+ * 1- Lexical rules input file is a text file.
+ * 2- Regular definitions are lines in the form LHS = RHS
+ * 3- Regular expressions are lines in the form LHS: RHS
+ * 4- Keywords are enclosed by { } in separate lines.
+ * 5- Punctuations are enclosed by [ ] in separate lines
+ * 6- \L represents Lambda symbol.
+ * 7- The following symbols are used in regular definitions and regular expressions with the meaning discussed in class: - | + * ( )
+ * 8- Any reserved symbol needed to be used within the language, is preceded by an escape backslash character.
+ * ***/
+vector<string> readInputFile(const string& filepath) {
+    ifstream file(filepath);
+
+    if (!file.is_open()) {
+        std::cerr << "Error opening the file.\n";
+        exit(1);
+    }
+
+    string line;
+    vector<string> lines;
+    while (getline(file, line)) {
+        lines.push_back(line);
+    }
+    file.close();
+    return lines;
+}
+
+vector<string> getRegularExpressions(const vector<string>& lines) {
+    vector<string> regularExpressions;
+    for (const auto& line : lines) {
+        if (line.find(':') != string::npos) {
```

Comment:
Let's consider that the given regular expression rules contain the following line:
"[{ } ; :]"
Now because ":" is present in that line, that line will be considered as a regular expression. However, it should be considered as a punctuation symbol.

Reply:
Yes, you are right we may solve that problem by checking that the first character in the line is not [ ,and that the last character is not ]